src/test/java/picard/TestDataProviders.java

```diff
@@ -21,21 +26,63 @@
 public class TestDataProviders {
 
     @Test
-    public void testAllDataProviders() throws IllegalAccessException, InstantiationException, InvocationTargetException {
-        int i = 0;
-        final ClassFinder classFinder = new ClassFinder();
-        classFinder.find("picard", Object.class);
-        for (Class testClass : classFinder.getClasses()) {
-            if (Modifier.isAbstract(testClass.getModifiers())) continue;
-            for (final Method method : testClass.getMethods()) {
-                if (method.isAnnotationPresent(DataProvider.class)) {
-                    System.err.println("Method: " + testClass.getName() + ":" + method.getName());
-                    method.invoke(testClass.newInstance());
-                    i++;
+    public void independentTestOfDataProviderTest() throws Exception {
+        final Iterator<Object[]> data = TestNGUtil.getDataProviders("picard");
+
+        Assert.assertTrue(data.hasNext(), "Found no data from testAllDataProvidersdata. Something is wrong.");
+
+        Assert.assertEquals( StreamSupport.stream(Spliterators.spliteratorUnknownSize(data, 0), false)
+                        .filter(c -> ((Method) c[0]).getName().equals("testAllDataProvidersdata")).count(), 1,
+                "getDataProviders didn't find testAllDataProvidersdata, which is in this class. Something is wrong.");
+    }
+
+    @DataProvider(name = "DataprovidersThatDontTestThemselves")
+    public Iterator<Object[]> testAllDataProvidersdata() throws Exception {
+        return TestNGUtil.getDataProviders("picard");
+    }
+
+    // runs all the @DataProviders it gets from DataprovidersThatDontTestThemselves.
+    // runs the BeforeSuite and BeforeClass methods of the same class before it runs the provider itself.
+
+    // @NoInjection annotations required according to this test:
+    // https://github.com/cbeust/testng/blob/master/src/test/java/test/inject/NoInjectionTest.java
+    @Test(dataProvider = "DataprovidersThatDontTestThemselves")
+    public void testDataProviderswithDP(@NoInjection final Method method, final Class clazz) throws
+            IllegalAccessException, InstantiationException {
+
+        Object instance = clazz.newInstance();
+
+        Set<Method> methodSet = Sets.newHashSet();
+        methodSet.addAll(Arrays.asList(clazz.getDeclaredMethods()));
+        methodSet.addAll(Arrays.asList(clazz.getMethods()));
+
+        // Some tests assume that the @BeforeSuite methods will be called before the @DataProviders
```

Comment:
This seems like a problem. This can (will always?) result in `BeforeSuite` code being called more than once for a test suite. That could be a problem, as BeforeSuite code can assume it will only be called once.

Reply:
OK. I'll see if I can leave out classes that have a @BeforeSuite and not call it.

Reply:
done.